pkg/controller/deployment/util/deployment_util.go

```diff
@@ -691,6 +800,69 @@ func IsRollingUpdate(deployment *extensions.Deployment) bool {
 	return deployment.Spec.Strategy.Type == extensions.RollingUpdateDeploymentStrategyType
 }
 
+// IsDeploymentComplete considers a deployment as complete once its desired replicas equals its
+// updatedReplicas and it doesn't violate minimum availability.
+func IsDeploymentComplete(deployment *extensions.Deployment, newStatus extensions.DeploymentStatus) bool {
+	return newStatus.UpdatedReplicas == deployment.Spec.Replicas &&
+		newStatus.AvailableReplicas >= deployment.Spec.Replicas-MaxUnavailable(*deployment)
+}
+
+// IsDeploymentProgressing reports if a deployment is progressing.
+func IsDeploymentProgressing(deployment *extensions.Deployment, newStatus extensions.DeploymentStatus) bool {
+	return deployment.Status.Replicas != newStatus.Replicas ||
+		deployment.Status.UpdatedReplicas != newStatus.UpdatedReplicas ||
+		deployment.Status.AvailableReplicas != newStatus.AvailableReplicas ||
+		deployment.Status.UnavailableReplicas != newStatus.UnavailableReplicas
```

Comment:
Updated `syncDeploymentStatus` to be re-entrant. @mfojtik even though we discussed about moving all the replica set annotation updates to a different PR (ideally the one that adds automatic rollback), I think the work there will be the same. I also want feedback from @bgrant0607, @pwittrock, and @janetkuo for the current state of this PR.